tests/arch/arc/interrupt-controller/src/interrupt-controller.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2021 Synopsys, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <zephyr.h>
+#include <sys/printk.h>
+#include <soc.h>
+
+#define TEST_IRQ_0_PRIVATE	22
+#define TEST_IRQ_1_PRIVATE	17
+#define TEST_IRQ_2_PRIVATE	18
+
+#define TEST_IRQ_0_SHARED	26
+#define TEST_IRQ_1_SHARED	27
+#define TEST_IRQ_2_SHARED	28
+
+#ifndef IRQ_ICI
+#define IRQ_ICI 19
+#endif
+
+BUILD_ASSERT(CONFIG_NUM_IRQ_PRIO_LEVELS >= 2, "test requires at least 2 priority levels");
+BUILD_ASSERT(CONFIG_NUM_IRQS >= TEST_IRQ_2_SHARED, "not enough interrupt lines for test");
+
+#ifdef CONFIG_ARC_CONNECT
+/*
+ * In case of SMP system all private interrupts are expected to have
+ * interrupt number < ARC_CONNECT_IDU_IRQ_START
+ * and shared interrupts have
+ * interrupt number >= ARC_CONNECT_IDU_IRQ_START
+ */
+BUILD_ASSERT(IRQ_ICI < ARC_CONNECT_IDU_IRQ_START);
+BUILD_ASSERT(TEST_IRQ_0_PRIVATE < ARC_CONNECT_IDU_IRQ_START);
+BUILD_ASSERT(TEST_IRQ_1_PRIVATE < ARC_CONNECT_IDU_IRQ_START);
+BUILD_ASSERT(TEST_IRQ_2_PRIVATE < ARC_CONNECT_IDU_IRQ_START);
+
+BUILD_ASSERT(TEST_IRQ_0_SHARED >= ARC_CONNECT_IDU_IRQ_START);
+BUILD_ASSERT(TEST_IRQ_1_SHARED >= ARC_CONNECT_IDU_IRQ_START);
+BUILD_ASSERT(TEST_IRQ_2_SHARED >= ARC_CONNECT_IDU_IRQ_START);
+#endif /* CONFIG_ARC_CONNECT */
+
+extern char __arc_isr_static_start[];
+extern char __arc_isr_static_end[];
+
+/* TODO: add locking */
+static void intc_prio_check_single(uint32_t irq, uint32_t expected_prio, const char *reason)
+{
+	uint32_t prio;
+
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
+	prio = z_arc_v2_aux_reg_read(_ARC_V2_IRQ_PRIORITY);
+	prio &= _ARC_V2_INT_PRIO_MASK;
+
+	zassert_true(prio == expected_prio,
+		"readback priority unexpected irq %u, expected %u, got %u, check reason %s\n",
+		irq, expected_prio, prio, reason);
+}
+
+
+static void intc_irq_check_masking_dynamic_single(uint32_t irq)
+{
```

Comment:
Is testing via the get/set functions enough, or should there be some real interrupt triggering test to see if the get/set are implemented correctly?